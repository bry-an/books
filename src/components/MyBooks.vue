<template>
  <div>
    <h2>My Selected Books</h2>
    <v-data-table
      no-data-text="You don't have any books selected yet"
      :headers="headers"
      :items="books"
    >
      <template v-slot:items="props">
        <td class="text-xs-center">
          <v-icon @click="$emit('remove-item', props.item)" small>remove_circle</v-icon>
        </td>
        <td class="text-xs-left">{{props.item.title}}</td>
        <td class="text-xs-left">{{ props.item.author_name[0] }}</td>
        <td class="text-xs-left">{{ props.item.first_publish_year}}</td>
        <td class="text-xs-left pa-0">
          <v-btn class="mr-3" primary small @click="routeToViewBook(props.item)">More Info</v-btn>
        </td>
      </template>
    </v-data-table>
  </div>
</template>

<script>
export default {
  name: "MyBooks",
  data() {
    return {
      headers: [
        {
          text: "Remove",
          sortable: false
        },
        {
          text: "Title",
          value: "title"
        },
        {
          text: "Author",
          value: "author_name[0]"
        },
        {
          text: "First published",
          value: "first_publish_year"
        },
        {
          text: "More info"
        }
      ]
    };
  },
  props: {
    books: {
      type: Array
    }
  },
  methods: {
    routeToViewBook(book) {
      this.$router.push({ name: "BookView", params: book });
    }
  }
};
</script>

<style lang="scss" scoped>
</style>