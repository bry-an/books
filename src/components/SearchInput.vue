<template>
  <div>
    <h4>Search by author, title, or subject</h4>
    <v-form>
      <v-text-field label="Search" placeholder="e.g. Hemmingway" v-model="userSearch"></v-text-field>
    </v-form>
    <v-btn @click="submitSearch" id="search-submit-btn" color="primary">Search</v-btn>
    <!-- <button @click="submitSearch">Submit</button> -->
  </div>
</template>

<script>
import axios from "axios";
export default {
  name: "SearchInput",
  data() {
    return {
      userSearch: "",
      searchResults: [],
      numOfResults: null
    };
  },
  methods: {
    async submitSearch() {
      console.log("submit search!");
      const queryUrl = `http://openlibrary.org/search.json?q=${
        this.userSearch
      }`;
      const result = {
        results: [],
        numOfResults: 0
      };

      await axios.get(queryUrl).then(response => {
        result.results = [...response.data.docs];
        result.numOfResults = response.data.numFound;
      });
      this.$emit("set-results", result);
    }
  }
};
</script>

<style lang="sass" scoped>
input[type="text"]
    display: block
    width: 200px
    margin: 8px 0
    border: 0.5px solid
</style>