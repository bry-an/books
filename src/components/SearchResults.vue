<template>
  <div>
    <h2>Your search found {{results.numOfResults}} results</h2>
    <p v-if="results.numOfResults > 100">Here are the most relevant 100 results</p>
    <v-data-table :headers="headers" :items="booksArray">
      <template v-slot:items="props">
        <td>{{props.item.title}}</td>
        <td>{{ props.item.author_name[0] }}</td>
        <td>{{ props.item.first_publish_year}}</td>
        <td>
          <v-btn primary small @click="$emit('add-to-my-books', props.item)">Add to My Books</v-btn>
        </td>
      </template>
    </v-data-table>
  </div>
</template>

<script>
export default {
  name: "SearchResults",
  props: {
    results: {
      type: Object
    }
  },
  data() {
    return {
      headers: [
        {
          text: "Title",
          value: "title"
        },
        {
          text: "Author",
          value: "author_name[0]"
        },
        {
          text: "First published",
          value: "first_publish_year"
        },
        {
          text: "Add Book"
        }
      ]
    };
  },
  computed: {
    booksArray() {
      return this.results.results;
    }
  }
};
</script>

<style lang="scss" scoped>
</style>