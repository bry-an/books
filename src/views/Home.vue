<template>
  <v-container>
    <v-layout row>
      <v-flex fill-width>
        <div class="title mb-4">Welcome to Booj Books!</div>
        <SearchInput @set-results="setResults"/>
      </v-flex>
    </v-layout>
    <v-layout row>
      <v-flex md6>
        <div
          class="body-2"
        >When you add a book to My Selected Books, you can view more information about the book from that list.</div>
        <SearchResults v-if="results" @add-to-my-books="addToMyBooks" :results="results"/>
      </v-flex>
      <v-flex md6 class="ml-5">
        <MyBooks :books="myBooks" @remove-item="removeFromMyBooks"/>
      </v-flex>
    </v-layout>
  </v-container>
</template>

<script>
import SearchInput from "../components/SearchInput";
import SearchResults from "../components/SearchResults";
import MyBooks from "../components/MyBooks";

export default {
  name: "Home",
  components: {
    SearchInput,
    MyBooks,
    SearchResults
  },
  data() {
    return {
      results: null,
      myBooks: []
    };
  },
  methods: {
    setResults(results) {
      this.results = results;
    },
    addToMyBooks(book) {
      this.myBooks.push(book);
    },
    removeFromMyBooks(book) {
      console.log("remove book book", book);
      this.myBooks = this.myBooks.filter(myBook => myBook.key !== book.key);
    }
  }
};
</script>
