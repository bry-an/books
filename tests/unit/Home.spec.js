import { shallowMount } from "@vue/test-utils"
import Vuetify from "vuetify"
import Home from "../../src/views/Home.vue"
import MyBooks from "../../src/components/MyBooks.vue"
import SearchInput from "../../src/components/SearchInput.vue"
import SearchResults from "../../src/components/SearchResults.vue"
import Vue from "vue"
Vue.use(Vuetify)

describe("Home", () => {
    let build
    build = () => {
        const books = [{
            title: "",
            author_name: [""],
            first_publish_year: 0
        }]
        const wrapper = shallowMount(Home)
        wrapper.setData({
            myBooks: books,
            results: { books }
        })
        return { wrapper }
    }

    it("renders the component", () => {
        const { wrapper } = build()
        expect(wrapper.html()).toMatchSnapshot()
    })
    it("contains search input, results, and mybooks components", () => {
        const { wrapper } = build()
        expect(wrapper.contains(MyBooks)).toBe(true)
        expect(wrapper.contains(SearchInput)).toBe(true)
        expect(wrapper.contains(SearchResults)).toBe(true)
    })
})