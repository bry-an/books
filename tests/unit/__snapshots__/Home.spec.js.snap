// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders the component 1`] = `
<v-container-stub tag="div">
  <v-layout-stub tag="div" row="">
    <v-flex-stub tag="div" fill-width="">
      <div class="title mb-4">Welcome to Booj Books!</div>
      <searchinput-stub></searchinput-stub>
    </v-flex-stub>
  </v-layout-stub>
  <v-layout-stub tag="div" row="">
    <v-flex-stub tag="div" md6="">
      <div class="body-2">When you add a book to My Selected Books, you can view more information about the book from that list.</div>
      <searchresults-stub results="[object Object]"></searchresults-stub>
    </v-flex-stub>
    <v-flex-stub tag="div" md6="" class="ml-5">
      <mybooks-stub books="[object Object]"></mybooks-stub>
    </v-flex-stub>
  </v-layout-stub>
</v-container-stub>
`;
